src/Common/Core/Impl/Services/ServiceManager.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Common.Core.Diagnostics;
+
+namespace Microsoft.Common.Core.Services {
+    public class ServiceManager : IServiceManager {
+        class ServiceFactory {
+            public IServiceFactory Factory { get; set; }
+            public object[] Parameters { get; set; }
+        }
+
+        private readonly Dictionary<Type, object> _services = new Dictionary<Type, object>();
+        private readonly Dictionary<Type, ServiceFactory> _deferredServices = new Dictionary<Type, ServiceFactory>();
```

Comment:
I'm not sure I get it. Even with current signature of `AddService(Type type, IServiceFactory factory = null, params object[] parameters)` you can do it like `new Func<object>(() => factory(parameters))`

---
In reply to: [106226285](https://github.com/Microsoft/RTVS/pull/3302#discussion_r106226285) [](ancestors = 106226285)